notebooks/04_e2e.ipynb

```diff
@@ -0,0 +1,578 @@
+    "    AdvectionDiffusion as AdvectionDiffusionMultichannel,\n",
+    ")\n",
+    "from autocast.data.datamodule import SpatioTemporalDataModule, TheWellDataModule\n",
+    "from autocast.metrics.spatiotemporal import MAE, MSE, RMSE\n",
+    "\n",
+    "THE_WELL = False\n",
+    "n_steps_input = 1\n",
+    "n_steps_output = 4\n",
+    "stride = 1\n",
+    "rollout_stride = 4\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "3",
+   "metadata": {},
+   "source": [
+    "### Read combined data into datamodule\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "4",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "if not THE_WELL:\n",
+    "    simulation_name = \"reaction_diffusion\"\n",
+    "    # simulation_name = \"advection_diffusion\"\n",
+    "    # simulation_name = \"advection_diffusion_multichannel\"\n",
+    "\n",
+    "    if simulation_name == \"advection_diffusion_multichannel\":\n",
+    "        # Override to use multichannel version\n",
+    "        Sim = AdvectionDiffusionMultichannel\n",
+    "    if simulation_name == \"reaction_diffusion\":\n",
+    "        Sim = ReactionDiffusion\n",
+    "    if simulation_name == \"advection_diffusion\":\n",
+    "        Sim = AdvectionDiffusion\n",
+    "\n",
+    "    sim = Sim(return_timeseries=True, log_level=\"error\")\n",
+    "\n",
+    "    def generate_split(\n",
+    "        simulator, n_train: int = 200, n_valid: int = 20, n_test: int = 20\n",
+    "    ):\n",
+    "        \"\"\"Generate training, validation, and test splits from the simulator.\"\"\"\n",
+    "        train = simulator.forward_samples_spatiotemporal(n_train)\n",
+    "        valid = simulator.forward_samples_spatiotemporal(n_valid)\n",
+    "        test = simulator.forward_samples_spatiotemporal(n_test)\n",
+    "        return {\"train\": train, \"valid\": valid, \"test\": test}\n",
+    "\n",
+    "    # Cache file path\n",
+    "    cache_file = Path(f\"{simulation_name}_cache.pkl\")\n",
+    "\n",
+    "    # Load from cache if it exists, otherwise generate and save\n",
+    "    if cache_file.exists():\n",
+    "        print(f\"Loading cached simulation data from {cache_file}\")\n",
+    "        with open(cache_file, \"rb\") as f:\n",
+    "            combined_data = pickle.load(f)\n",
+    "            for key in [\"data\", \"constant_scalars\", \"constant_fields\"]:\n",
+    "                combined_data[\"test\"][key] = (\n",
+    "                    combined_data[\"test\"][key][:5]\n",
+    "                    if combined_data[\"test\"][key] is not None\n",
+    "                    else None\n",
+    "                )\n",
+    "    else:\n",
+    "        print(\"Generating simulation data...\")\n",
+    "        combined_data = generate_split(sim)\n",
+    "        print(f\"Saving simulation data to {cache_file}\")\n",
+    "        with open(cache_file, \"wb\") as f:\n",
+    "            pickle.dump(combined_data, f)\n",
```

Comment:
the data modules expect there to be separate train/val/test files

Reply:
I like the idea of capturing the split logical to make the dataset so it can be read this way.

The idea origianlly was to have a more temporary cache of a simulation available for prototyping without needing to change the datamodule args or modify a more permanent data source in `datasets/` that's being read in.

Perhaps we could write to `data_path = f"../datasets/{simulation_name}_tmp/" as an indicator that outputs from the notebook are intended to be more temporary cache locations?

Reply:
I've updated in [7ba4bf3](https://github.com/alan-turing-institute/autocast/pull/117/commits/7ba4bf38c1de3fc4e37ed2ddea5563bfd105129f) with the first option and to write to `datasets/tmp/simulation_name` and added a util function to `get_datamodule` so this can be used elsewhere.